File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/ios/Classes/Model/IDOHandWashingReminderParamModel.swift
//
//  IDOHandWashingReminderParamModel.swift
//  protocol_channel
//
//  Created by hc on 2023/10/12.
//

import Foundation
/// Set Hand Washing Reminder Event
public class IDOHandWashingReminderParamModel: NSObject, IDOBaseModel {
    /// 0: Off
    /// 1: On
    /// Default is off
    public var onOff: Int
    /// Start hour of the reminder
    public var startHour: Int
    /// Start minute of the reminder
    public var startMinute: Int
    /// End hour of the reminder
    public var endHour: Int
    /// End minute of the reminder
    public var endMinute: Int
    /// Repeat
    public var repeats: Set<IDOWeek>
    /// Reminder interval in minutes
    /// Default is 60 minutes
    public var interval: Int

    enum CodingKeys: String, CodingKey {
        case onOff = "on_off"
        case startHour = "start_hour"
        case startMinute = "start_minute"
        case endHour = "end_hour"
        case endMinute = "end_minute"
        case repeats = "repeat"
        case interval = "interval"
    }

    public init(onOff: Int, startHour: Int, startMinute: Int, endHour: Int, endMinute: Int, repeats: Set<IDOWeek>, interval: Int) {
        self.onOff = onOff
        self.startHour = startHour
        self.startMinute = startMinute
        self.endHour = endHour
        self.endMinute = endMinute
        self.repeats = repeats
        self.interval = interval
    }
    
    required public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        onOff = try container.decode(Int.self, forKey: .onOff)
        startHour = try container.decode(Int.self, forKey: .startHour)
        startMinute = try container.decode(Int.self, forKey: .startMinute)
        endHour = try container.decode(Int.self, forKey: .endHour)
        endMinute = try container.decode(Int.self, forKey: .endMinute)
        interval = try container.decode(Int.self, forKey: .interval)

        if let value = try container.decodeIfPresent(Int.self, forKey: .repeats) {
            var items = Set<IDOWeek>()
            for i in 0...6 {
                if value & (1 << (i + 1)) != 0 {
                    items.insert(IDOWeek(rawValue: i)!)
                }
            }
            repeats = items
        } else {
            repeats = Set<IDOWeek>()
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(onOff, forKey: .onOff)
        try container.encode(startHour, forKey: .startHour)
        try container.encode(startMinute, forKey: .startMinute)
        try container.encode(endHour, forKey: .endHour)
        try container.encode(endMinute, forKey: .endMinute)
        try container.encode(interval, forKey: .interval)

        var value = 0
        for item in repeats {
            value |= (1 << (item.rawValue + 1))
        }
        try container.encode(value, forKey: .repeats)
    }

    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            let jsonString = String(data: jsonData, encoding: .utf8)
            return jsonString
        }
        return nil
    }
}

/// Set Hand Washing Reminder Event
@objcMembers
public class IDOHandWashingReminderParamModelObjc: NSObject, IDOBaseModel {
    /// 0: Off
    /// 1: On
    /// Default is off
    public var onOff: Int
    /// Start hour of the reminder
    public var startHour: Int
    /// Start minute of the reminder
    public var startMinute: Int
    /// End hour of the reminder
    public var endHour: Int
    /// End minute of the reminder
    public var endMinute: Int
    /// Repeat
    public var repeats: [IDOWeekObjc]
    /// Reminder interval in minutes
    /// Default is 60 minutes
    public var interval: Int

    enum CodingKeys: String, CodingKey {
        case onOff = "on_off"
        case startHour = "start_hour"
        case startMinute = "start_minute"
        case endHour = "end_hour"
        case endMinute = "end_minute"
        case repeats = "repeat"
        case interval = "interval"
    }

    public init(onOff: Int, startHour: Int, startMinute: Int, endHour: Int, endMinute: Int, repeats: [IDOWeekObjc], interval: Int) {
        self.onOff = onOff
        self.startHour = startHour
        self.startMinute = startMinute
        self.endHour = endHour
        self.endMinute = endMinute
        self.repeats = repeats
        self.interval = interval
    }
    
    required public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        onOff = try container.decode(Int.self, forKey: .onOff)
        startHour = try container.decode(Int.self, forKey: .startHour)
        startMinute = try container.decode(Int.self, forKey: .startMinute)
        endHour = try container.decode(Int.self, forKey: .endHour)
        endMinute = try container.decode(Int.self, forKey: .endMinute)
        interval = try container.decode(Int.self, forKey: .interval)

        if let value = try container.decodeIfPresent(Int.self, forKey: .repeats) {
            var items = [IDOWeekObjc]()
            for i in 0...6 {
                if ((value >> i) & 1) > 0 {
                    items.append(IDOWeekObjc(week: IDOWeek(rawValue: i)!))
                }
            }
            repeats = items
        } else {
            repeats = [IDOWeekObjc]()
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(onOff, forKey: .onOff)
        try container.encode(startHour, forKey: .startHour)
        try container.encode(startMinute, forKey: .startMinute)
        try container.encode(endHour, forKey: .endHour)
        try container.encode(endMinute, forKey: .endMinute)
        try container.encode(interval, forKey: .interval)

        var value = 0
        for item in repeats {
            value |= (1 << (item.week.rawValue + 1))
        }
        try container.encode(value, forKey: .repeats)
    }

    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            let jsonString = String(data: jsonData, encoding: .utf8)
            return jsonString
        }
        return nil
    }
}

File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/android/src/main/kotlin/com/idosmart/model/IDOHandWashingReminderParamModel.kt
//
//  IDOHandWashingReminderParamModel.kt
//  protocol_channel
//
//  Created by hedongyang on 2023/10/30.
//

package com.idosmart.model

import com.google.gson.*
import java.lang.reflect.Type

///
/// Set Hand Washing Reminder Event

open class IDOHandWashingReminderParamModel(
    onOff: Int,
    startHour: Int,
    startMinute: Int,
    endHour: Int,
    endMinute: Int,
    repeats: HashSet<IDOWeek>,
    interval: Int
) : IDOBaseAdapterModel<IDOHandWashingReminderParamModel> {

    /// 0: Off
    /// 1: On
    /// Default is off
    var onOff: Int = onOff

    /// Start hour of the reminder
    var startHour: Int = startHour

    /// Start minute of the reminder
    var startMinute: Int = startMinute

    /// End hour of the reminder
    var endHour: Int = endHour

    /// End minute of the reminder
    var endMinute: Int = endMinute

    /// Repeat
    var repeats: HashSet<IDOWeek> = repeats

    /// Reminder interval in minutes
    /// Default is 60 minutes
    var interval: Int = interval
    override fun getDeSerializer(): IDOBaseModelDeSerializer<IDOHandWashingReminderParamModel>? {
        return HandWashingReminderParamModelSerializer()
    }

    override fun toString(): String {
        return GsonBuilder().create().toJson(this).toString()
    }
}


internal class HandWashingReminderParamModelSerializer : IDOBaseModelDeSerializer<IDOHandWashingReminderParamModel> {
    override fun serialize(src: IDOHandWashingReminderParamModel, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val jsonObject = JsonObject()
        jsonObject.addProperty("on_off", src.onOff)
        jsonObject.addProperty("start_hour", src.startHour)
        jsonObject.addProperty("start_minute", src.startMinute)
        jsonObject.addProperty("end_hour", src.endHour)
        jsonObject.addProperty("end_minute", src.endMinute)
        jsonObject.addProperty("interval", src.interval)

        var num = 0
        src.repeats.forEach { week ->
            num = num or (1 shl week.rawValue + 1)
        }
        jsonObject.addProperty("repeat", num)
        return jsonObject
    }

    override fun deserialize(
        json: JsonElement?,
        typeOfT: Type?,
        context: JsonDeserializationContext?
    ): IDOHandWashingReminderParamModel? {
        json?.let {
            val jsonObject = it.asJsonObject
            val onOff = jsonObject.get("on_off").asInt
            val startHour = jsonObject.get("start_hour").asInt
            val startMinute = jsonObject.get("start_minute").asInt
            val endHour = jsonObject.get("end_hour").asInt
            val endMinute = jsonObject.get("end_minute").asInt
            val interval = jsonObject.get("interval").asInt

            val repeat = jsonObject.get("repeat").asInt
            val items = HashSet<IDOWeek>()
            for (i in 0..6) {
                if (repeat and (1 shl (i + 1)) != 0) {
                    items.add(IDOWeek.fromRawValue(i))
                }
            }

            return IDOHandWashingReminderParamModel(
                onOff,
                startHour,
                startMinute,
                endHour,
                endMinute,
                items,
                interval
            )
        }
        return null
    }
}